File: apps/mobile/lib/utils/transcript_timing.dart
class TranscriptSegment {
  const TranscriptSegment({
    required this.text,
    required this.start,
    required this.end,
  });

  final String text;
  final Duration start;
  final Duration end;
}

List<TranscriptSegment> estimateTranscriptTiming(
  String? script,
  Duration audioDuration,
) {
  final trimmedScript = script?.trim();
  if (trimmedScript == null || trimmedScript.isEmpty) {
    return const [];
  }

  final rawParagraphs = trimmedScript
      .split(RegExp(r'\n\s*\n+'))
      .map((paragraph) => paragraph.trim())
      .where((paragraph) => paragraph.isNotEmpty)
      .toList(growable: false);

  final texts =
      rawParagraphs.length > 1 ? rawParagraphs : _splitSentences(trimmedScript);
  if (texts.isEmpty) return const [];

  if (audioDuration <= Duration.zero) {
    return [
      for (final text in texts)
        TranscriptSegment(text: text, start: Duration.zero, end: Duration.zero),
    ];
  }

  final weights = texts.map(_timingWeight).toList(growable: false);
  final totalWeight = weights.fold<int>(0, (total, weight) => total + weight);
  final durationMs = audioDuration.inMilliseconds;
  var accumulatedWeight = 0;

  return [
    for (var index = 0; index < texts.length; index++)
      TranscriptSegment(
        text: texts[index],
        start: _scaledDuration(accumulatedWeight, totalWeight, durationMs),
        end: index == texts.length - 1
            ? audioDuration
            : _scaledDuration(
                accumulatedWeight += weights[index],
                totalWeight,
                durationMs,
              ),
      ),
  ];
}

List<String> _splitSentences(String script) {
  final matches = RegExp(r'[^。！？.!?]+[。！？.!?]?')
      .allMatches(script)
      .map((match) => match.group(0)?.trim() ?? '')
      .where((text) => text.isNotEmpty)
      .toList(growable: false);

  return matches.isEmpty ? [script] : matches;
}

int _timingWeight(String text) {
  final compact = text.replaceAll(RegExp(r'\s+'), '');
  return compact.runes.isEmpty ? 1 : compact.runes.length;
}

Duration _scaledDuration(int weight, int totalWeight, int durationMs) {
  if (totalWeight <= 0 || durationMs <= 0) return Duration.zero;
  return Duration(milliseconds: (weight / totalWeight * durationMs).round());
}

File: apps/app/src/components/podcast/transcriptTiming.ts
export interface TranscriptSegment {
  text: string;
  start: number;
  end: number;
}

export function estimateTranscriptTiming(
  script: string | null,
  audioDurationSeconds: number,
): TranscriptSegment[] {
  const trimmedScript = script?.trim();
  if (trimmedScript === undefined || trimmedScript === '') return [];

  const rawParagraphs = trimmedScript
    .split(/\n\s*\n+/)
    .map((paragraph) => paragraph.trim())
    .filter((paragraph) => paragraph !== '');
  const texts =
    rawParagraphs.length > 1 ? rawParagraphs : splitSentences(trimmedScript);
  if (texts.length === 0) return [];

  const duration = Number.isFinite(audioDurationSeconds)
    ? Math.max(0, audioDurationSeconds)
    : 0;
  if (duration <= 0) {
    return texts.map((text) => ({ text, start: 0, end: 0 }));
  }

  const weights = texts.map(timingWeight);
  const totalWeight = weights.reduce((total, weight) => total + weight, 0);
  let accumulatedWeight = 0;

  return texts.map((text, index) => {
    const start = scaledSeconds(accumulatedWeight, totalWeight, duration);
    accumulatedWeight += weights[index] ?? 0;
    const end =
      index === texts.length - 1
        ? duration
        : scaledSeconds(accumulatedWeight, totalWeight, duration);
    return { text, start, end };
  });
}

function splitSentences(script: string): string[] {
  const matches = script.match(/[^。！？.!?]+[。！？.!?]?/g);
  if (matches === null) return [script];
  const sentences = matches
    .map((match) => match.trim())
    .filter((match) => match !== '');
  return sentences.length > 0 ? sentences : [script];
}

function timingWeight(text: string): number {
  const compact = text.replace(/\s+/g, '');
  return compact.length > 0 ? Array.from(compact).length : 1;
}

function scaledSeconds(
  weight: number,
  totalWeight: number,
  duration: number,
): number {
  if (totalWeight <= 0 || duration <= 0) return 0;
  return (weight / totalWeight) * duration;
}
